### ISSUE TITLE
MIC calculation on outgoing AS2 is incorrect when using compress before signing

### ISSUE THREAD

**Issue by jthompsonnp** [phax/as2-lib#140]
Using as2lib as both client and server. When sending a non-compressed, or a compressed after signed file, the MIC for the MDN is matched as **MIC: Xc9uKhufeGS/tEPvxjbe9g==, md5**

But when using compress before sign, the original MIC is **MIC: bMUIBjelUf5Zy+kPrtaMRw==, md5**. And the MIC for the MDN is **MIC: Xc9uKhufeGS/tEPvxjbe9g==, md5**

**MIC IS NOT MATCHED; original MIC: bMUIBjelUf5Zy+kPrtaMRw==, md5 received MIC: Xc9uKhufeGS/tEPvxjbe9g==, md5 [<local-20220802-115713-0300-4490@local_ec2Key>]**

It seems that within the secure method of the AS2SenderModule, the data is cleared, and the original MIC is calculated with blank data as an input. However the actual as2 message that is sent has the correct data and thus the MDN is sent with the 'correct' MIC

**Comment by phax** [comment]
I think I understand the problem.
Officially the AS2 RFC defines a way how to do things (compress before sign or sign before compress). I need to check which one is correct. The other solution is just a workaround per se... I need to check

**Comment by jthompsonnp** [comment]
Compress before signing seems to be the default. But this is the only one that I am having trouble with.

![image](https://user-images.githubusercontent.com/46717211/182947055-da35cf32-71db-4c03-9b0e-afb9d6d1f750.png)

**Comment by phax** [comment]
Ok. Quote from RFC 5402, chapter 3
>    When compressing a document that will be signed, the application MAY compress the innermost MIME body before signing (see Sections 3.2 and 3.5), or it MAY compress the outer multipart/signed MIME body (see Sections 3.3 and 3.6), but it MUST NOT do both within the same document.  The receiving application MUST support both methods of compression when unpackaging an inbound document.

And chapter 4.1.  MIC Calculation for Signed Message
>   For any signed message, the MIC to be returned is calculated over the same data that was signed in the original message as per [AS1].  The signed content will be a MIME bodypart that contains either compressed or uncompressed data.

**Comment by jthompsonnp** [comment]
So it should work with **compress BEFORE sign** and with **compress AFTER sign**. But I'm getting a "MIC IS NOT MATCHED; " error whenever I test compress before sign.

Is there a reason why this is happening?
1. No compression - Original MIC and Returned MIC are matched as MIC: Xc9uKhufeGS/tEPvxjbe9g==, md5
2. Compress after sign - Original MIC and Returned MIC are matched as MIC: Xc9uKhufeGS/tEPvxjbe9g==, md5
3. Compress before sign - MIC IS NOT MATCHED; original MIC: bMUIBjelUf5Zy+kPrtaMRw==, md5 received MIC: Xc9uKhufeGS/tEPvxjbe9g==, md5

Even though the data I'm transferring is the same with all tests?

**Comment by jthompsonnp** [comment]
To add to this, when testing **_AS2Lib -> Mendelson_**. This is what I get:
1. No compression ------- **MIC is matched, MIC: Xc9uKhufeGS/tEPvxjbe9g==, md5**
2. Compress after sign --- **MIC is matched, MIC: Xc9uKhufeGS/tEPvxjbe9g==, md5**
3. Compress before sign - **MIC is matched, MIC: bMUIBjelUf5Zy+kPrtaMRw==, md5** 

So it looks like the MIC is calculated the same with **No Compression** as it is with **Compress After Sign**. But with **Compress Before Sign**, the MIC is calculated differently resulting in a different MIC.


However when testing **_AS2Lib -> AS2Lib_**, the received MIC for **Compress Before Sign** is the same as the original MIC for **No Compression** and **Compress After Sign**. Which makes me believe that it is being calculated incorrectly.

**Comment by phax** [comment]
Thanks for your thorough analysis 👍🏼 
I just currently don't have the immediate capacity to process your issue. But you provided all the relevant information!

**Comment by phax** [comment]
I checked with the latest version (4.11.0-SNAPSHOT) with my `MainSendToMendelsonTestServerPlayground` example application, using MD5 as the signing algo and `3des` as the (insecure) encryption algorithm:

1. No compression: `NcyVevXfkQbXw9ydKLxcyw==, md5`
2. Compress after sign: `NcyVevXfkQbXw9ydKLxcyw==, md5`
3. Compress before sign: `otv2eEXDRF2gW2xMPme5IA==, md5`

In all 3 cases the MIC is matched by Mendelson. So my assumption is, that the sending MIC calculation works as expected.

But as you mentioned the `as2lib-as2lib` sending has issues, so maybe I should focus on the MIC calculation for incoming message.

**Comment by jthompsonnp** [comment]
Yes, I believe that's where the issue is occurring. 

Thank you for taking a look at this issue. Let me know if you need any more information from me to recreate it.

**Comment by CarolF82** [comment]
This issue may because of the inconsistent logic of MIC calculation between AS2SenderModule and AS2ReceiverHandler. 
When Compress before sign is selected on sending, the MIC is calculated on the compressed MIME body. But in AS2ReceiverHanlder, the MIC calculation happens at the time of creating MDN message, which is after decompress is called. No matter which compression/sign setting on the sender side, the MIC is always calculated on the MIME body without compression. 
 
If possible, please reopen this issue, seems worth fixing it. Thanks

**Comment by phax** [comment]
@CarolF82  thanks for pointing me to the right spot - the problem is not the algorithm itself, but the place in the code where it is calculated...

**Comment by alvarolivie** [comment]
I think it may be related but not sure, I've been using the library for a couple of days and everything works great except that for inbound calls that are signed and/or encrypted (I was testing using mendelson) I keep getting a MIC mismatch. I fixed it using a local copy of the library and calculating the MIC only on the content and content headers while excluding boundaries. Not sure if this is expected or something to be fixed so just jotting this down here. If this is a fix you want to add I'm more than happy to make a PR with my changes.

**Comment by phax** [comment]
@alvarolivie  yes please make a PR - I already spent countless hours on this and I don't get it...
